### PR TITLE
DataStore feature for React Native 

### PR DESCRIPTION
**Pull request by Ashish-Nanda**
#### This PR adds the code to implement and enable the DataStore feature for React Native apps.
_Issue #, if available:_
Addresses feature request #4527  

_Description of changes:_

- Wrote a React Native specific storage adapter that uses AsyncStorage under the hood
- Created a database-like abstraction for AsyncStorage to make it easier to plug into the adapter
- To implement the notion of separate stores in AsyncStorage, we keep track of a collection of keys prefixed by storeName
- Logic to get default adapter now lives in platform specific files
- Wrote a React Native specific implementation for Reachability using @react-native-community/netinfo 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -22,7 +22,8 @@
 		"publish:master": "lerna publish --canary --force-publish \"*\" --yes --dist-tag=unstable --preid=unstable --exact",
 		"publish:beta": "lerna publish --canary --force-publish \"*\" --yes --dist-tag=beta --preid=beta --exact",
 		"publish:release": "lerna publish --conventional-commits --yes --message 'chore(release): Publish [ci skip]'",
-		"publish:1.0-stable": "lerna publish --conventional-commits --yes --dist-tag=stable-1.0 --message 'chore(release): Publish [ci skip]'"
+		"publish:1.0-stable": "lerna publish --conventional-commits --yes --dist-tag=stable-1.0 --message 'chore(release): Publish [ci skip]'",
+		"publish:rn-datastore": "lerna publish --canary --force-publish \"*\" --yes --dist-tag=rn-datastore --preid=rn-datastore --exact"
```
Comment (manueliglesias) [inline package.json]: Should we remove this before merging to master?

File: .circleci/config.yml [hunk 2]
```diff
@@ -336,6 +336,7 @@ workflows:
                 - master
                 - beta
                 - 1.0-stable
+                - rn-datastore
```
Comment (manueliglesias) [inline .circleci/config.yml]: Should we remove this before merging to master?

File: package.json (hunk 1 above)
Comment (Ashish-Nanda) [reply package.json]: Agree. I've removed it.

File: .circleci/config.yml (hunk 2 above)
Comment (Ashish-Nanda) [reply .circleci/config.yml]: Removed.

**Pull request merged**